### PR TITLE
Fixed #209 and #213 : fam_progress multithreaded tests

### LINKED ISSUE #209: Add example code for multi threaded model
Add an example code for using multi threaded model

### PR DESCRIPTION
**Pull request by cynara-justine**
Fixed https://github.com/OpenFAM/OpenFAM/issues/209 and  https://github.com/OpenFAM/OpenFAM/issues/213

### CODE REVIEW

File: test/unit-test/fam-api/fam_progress_mt.cpp [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+    pthread_barrier_wait(&barrier);
+    try {
+        for (int j = 0; j < NUM_IO_ITERATIONS; j++) {
+            int *local1 = (int *)malloc(1000000);
+            int *local2 = (int *)malloc(1000000);
+            ctx[i]->fam_get_nonblocking(local1, item, 500, (i + 1) * 100000);
+            ctx[i]->fam_get_nonblocking(local2, item, 10, (i + 1) * 100000);
+        }
+        ctx[i]->fam_quiet();
```
Comment (mashoodk) [inline test/unit-test/fam-api/fam_progress_mt.cpp:104]: Here, we may end up calling fam_quiet from multiple in same context. Do you want to consider changing this logic ?

File: test/unit-test/fam-api/fam_progress_mt.cpp [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+            ctx[i] = my_fam->fam_context_open();
+        }
+    } catch (Fam_Exception &e) {
+        printf("Create region failed: %d: %s\n", e.fam_error(),
+               e.fam_error_msg());
+        return -1;
+    }
+    if (pthread_barrier_init(&barrier, NULL, TOTAL_NUM_THREADS) != 0) {
+
+        cout << "Error in barrier initialization" << endl;
```
Comment (mashoodk) [inline test/unit-test/fam-api/fam_progress_mt.cpp:175]: extra spacing above this line.

File: test/unit-test/fam-api/fam_progress_mt.cpp [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+        info[i].tid = i;
+        if ((rc = pthread_create(&thr[i], NULL,
+                                 thrd_fam_context_mt_with_fam_progress,
+                                 &info[i]))) {
+            fprintf(stderr, "error: pthread_create, rc: %d\n", rc);
+            exit(1);
+        }
+    }
+
+    pthread_barrier_destroy(&barrier);
```
Comment (mashoodk) [inline test/unit-test/fam-api/fam_progress_mt.cpp:196]: Shouldn't pthread_barrier_destroy be called after pthread_join.

**Review by mashoodk — CHANGES_REQUESTED** [review]

File: test/unit-test/fam-api/fam_progress_mt.cpp (hunk 2 above)
Comment (cynara-justine) [reply test/unit-test/fam-api/fam_progress_mt.cpp:175]: Thanks, removed the extra line.

File: test/unit-test/fam-api/fam_progress_mt.cpp [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+using namespace std;
+using namespace openfam;
+fam *my_fam;
+Fam_Options fam_opts;
+
+Fam_Region_Descriptor *testRegionDesc;
+const char *testRegionStr;
+int rc;
+
+#define NUM_THREADS 32
```
[the remark this answers is not in the document]
Comment (cynara-justine) [reply test/unit-test/fam-api/fam_progress_mt.cpp:53]: With 50 iterations, it is completing in less than a second - now fam progress values come around 96, 2 etc. then 0. So the functionality is working and so we decided to proceed.
With 500 iterations, fam_progress values are in the range of 2500.  As it takes a long time, we have set the number of iterations to 50 for the test program that is being checked in

File: test/unit-test/fam-api/fam_progress_mt.cpp [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+typedef struct {
+    Fam_Descriptor *item;
+    uint64_t offset;
+    int32_t tid;
+    int32_t deltaValue;
+} ValueInfo;
+
+void *thr_check_fam_progress(void *arg) {
+    uint64_t progress = 0;
+    cpu_set_t cpuset;
```
Comment (mashoodk) [inline test/unit-test/fam-api/fam_progress_mt.cpp:71]: You can comment the cpu set affinity code.

File: test/unit-test/fam-api/fam_progress_mt.cpp [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+    sched_setaffinity(0, sizeof(cpuset), &cpuset);
+
+    pthread_barrier_wait(&barrier);
+    while (activethreads != 0) {
+        for (int i = 0; i < NUM_CONTEXTS; i++) {
+            try {
+                progress = ctx[i]->fam_progress();
+                pthread_yield();
+                if (progress != 0) {
+                    cout << "I/Os in Progress for context" << i << " is "
```
Comment (mashoodk) [inline test/unit-test/fam-api/fam_progress_mt.cpp:95]: Instead of printing all the value, you can print maximum value.

**Review by mashoodk — CHANGES_REQUESTED** [review]

**Review by faizan-barmawer — CHANGES_REQUESTED** [review]
Overall tests look fine.
Few minor changes to handle.

Regards,
Faizan

**Review by mashoodk — APPROVED** [review]

**Review by faizan-barmawer — APPROVED** [review]
Code looks good to merge.

**Pull request merged**